sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -2912,6 +2912,32 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     RefreshTableStatement(visitMultipartIdentifier(ctx.multipartIdentifier()))
   }
 
+  /**
+   * A command for users to list the column names for a table.
+   * This function creates a [[ShowColumnsStatement]] logical plan.
+   *
+   * The syntax of using this command in SQL is:
+   * {{{
+   *   SHOW COLUMNS (FROM | IN) tableName=multipartIdentifier
+   *        ((FROM | IN) namespace=multipartIdentifier)?
+   * }}}
+   */
+  override def visitShowColumns(ctx: ShowColumnsContext): LogicalPlan = withOrigin(ctx) {
+    import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
+
+    val table = visitMultipartIdentifier(ctx.table)
+    val namespace = Option(ctx.namespace).map(visitMultipartIdentifier)
+    if (namespace.isDefined && namespace.get.length > 1) {
+      throw new ParseException(
+        s"Namespace name should have only one part if specified: ${namespace.get.quoted}", ctx)
+    }
+    if (table.length > 2) {
+      throw new ParseException(
+        s"Table name should have at most two parts: ${table.quoted}", ctx)
+    }
```

Comment:
Should we move the checks to Analyzer? Let ResolveSessionCatalog do the check?

Reply:
cc @cloud-fan

Reply:
I put this validations here because It's sounds like sintaxis problems, and I saw other ParserExceptions in AstBuilder. In ResolveSessionCatalog I saw AnalisisExceptions that sounds more like catalog access problems. 
But is only by intuition, we can put ir wherever you want. 
Could be better to put all the validations together

Reply:
This is a tricky problem. Ideally, we should have a check at the parser that: if namespace is specified, then table name should be a single part.

Unfortunately we support `SHOW COLUMNS FROM db1.t1 IN db1`. To keep supporting it, we can't check anything in the parser. In the analyzer, we can add a check before converting to the v1 command: if namespace is specified, it must be a single part, and the table name must have <= 2 parts.

Reply:
Ok, perfect, I move this validation